### PR TITLE
Prototyping the BQ to BT solution for Syndeo

### PR DESCRIPTION
**Pull request by pabloem**

### CODE REVIEW

File: syndeo-template/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+# Running locally
+mvn exec:java -Dexec.mainClass="com.google.cloud.syndeo.SyndeoTemplate" \
+  -Dexec.args="--runner=DataflowRunner --project=$GCP_PROJECT --region=$REGION \
+               --gcpTempLocation=$GCS_BUCKET --pipelineSpec=$PATH_TO_SPEC --tempLocation=$GCS_BUCKET/temp"
+
+# Running a pushed template (see Push template below)
+Run Template on Dataflow from Uploaded Template:
+gcloud dataflow flex-template run "syndeojob-`date +%Y%m%d-%H%M%S`" \
+  --template-file-gcs-location "gs://$GCS_BUCKET/template.json" --region $REGION \
+  --temp-location "gs://$GCS_BUCKET/temp" --parameters=pipelineSpec="gs://$GCS_BUCKET/pubsub_to_avro_config.txt"
```
Comment (damondouglas) [inline syndeo-template/README.md:84]: Thank you for making this available! Is `pubsub_to_avro` meant to be `bigquery_to_bigtable`?

Comment (pabloem) [reply syndeo-template/README.md:84]: ah correct!

**Comment by pabloem** [comment]
r: @zhoufek

**Comment by pabloem** [comment]
maybe r: @damondouglas as well since he's worked on a lot of these transforms : )

File: syndeo-template/README.md [hunk 2]
```diff
@@ -18,6 +18,50 @@ transforms. Then, the template will take the pipeline spec, and
...
+mvn -B spotless:apply compile -f  unified-templates.xml -pl syndeo-template/pom.xml
+```
+
+### Run tests (unit and integration tests)
+
+To **run unit tests** for the Syndeo template, run the following command. Note that this command knows to skip
+integration tests and only runs unit tests:
+
+```shell
+mvn clean package test -f unified-templates.xml -pl syndeo-template/pom.xml
```
Comment (damondouglas) [inline syndeo-template/README.md:35]: I ran this command in a Google Cloud Shell after cloning https://github.com/pabloem/DataflowTemplates/tree/bq-2-bt.

```
mvn clean package test -f unified-templates.xml -pl syndeo-template/pom.xml
```

I received the following error:

```
[ERROR] Failed to execute goal on project syndeo-template: Could not resolve dependencies for project com.google.cloud.teleport.syndeo:syndeo-template:jar:1.0-SNAPSHOT: Failure to find com.google.cloud.teleport:integration-testing-lib:jar:1.0-SNAPSHOT in https://repository.apache.org/content/repositories/snapshots was cached in the local repository, resolution will not be reattempted until the update interval of apache.snapshots has elapsed or updates are forced
```

File: syndeo-template/README.md [hunk 3]
```diff
@@ -18,6 +18,50 @@ transforms. Then, the template will take the pipeline spec, and
...
+```
+
+The file `BigTableWriteIT` **holds integration tests** for the basic BigTable Syndeo integration. These integration tests
+rely on the existence of a BigTable instance and table, as well as a BigQuery dataset, which holds the BQ data that
+is part of the pipeline's read.
+
```
Comment (damondouglas) [inline syndeo-template/README.md:41]: May we consider inserting the following?  I don't know if further instructions are needed.  I couldn't complete the instructions to see what else would be needed for setup due to the `Could not resolve dependencies for project com.google.cloud.teleport.syndeo:syndeo-template:jar:1.0-SNAPSHOT` error noted above. 

--------------

To set up your Google Cloud project for the integration tests, the following steps assume you have installed and setup [gcloud](https://cloud.google.com/sdk/gcloud).

1. Set your default project.

```shell
gcloud config set project PROJECT
```

2. Create the BigQuery dataset.

```
bq mk syndeo_dataset
```

3. Create the BigTable instance.

```
gcloud bigtable instances create teleport --display-name=teleport --cluster-config=id=teleport,zone=us-central1-a
```

4. Create the artifact bucket

```
ARTIFACT_BUCKET=[CHANGE ME]
gsutil mb gs://$ARTIFACT_BUCKET
```

5. Create the temporary location bucket

```
TEMP_LOCATION_BUCKET=[CHANGE ME]
gsutil mb gs://$TEMP_LOCATION_BUCKET
```

File: syndeo-template/README.md [hunk 4]
```diff
@@ -18,6 +18,50 @@ transforms. Then, the template will take the pipeline spec, and
...
+```shell
+mvn clean package test -f unified-templates.xml -pl syndeo-template/pom.xml  \
```
Comment (damondouglas) [inline syndeo-template/README.md:43]: Assuming the above README inserts:

--------------------------

```
mvn clean package test -f unified-templates.xml -pl syndeo-template/pom.xml  \
    -Dtest="BigTableWriteIT#testBigQueryToBigTableSmallNonTemplateJob"  \
    -Dproject="$(gcloud config get-value project)"  -DartifactBucket="gs://$ARTIFACT_BUCKET"  \
    -Dregion="us-central1" -DtempLocation=gs://$TEMP_LOCATION_BUCKET
```

**Review by damondouglas — CHANGES_REQUESTED** [review]
@pabloem

File: syndeo-template/README.md (hunk 3 above)
Comment (pabloem) [reply syndeo-template/README.md:41]: this is awesome. Thanks!

File: syndeo-template/README.md (hunk 2 above)
Comment (pabloem) [reply syndeo-template/README.md:35]: this means you'll need to build the integration test project first? Have you tried that? (I don't know how to make sure it gets built first...)

**Review by damondouglas — CHANGES_REQUESTED** [review]
@pabloem When I ran the command `mvn compile exec:java -Dexec.mainClass="com.google.cloud.syndeo.ConfigGen"`, I received the error below.  Feel free to merge this if you think the error is irrelevant.

```
java.lang.ClassNotFoundException: com.google.cloud.syndeo.WritePipelineSpecForTesting
    at java.net.URLClassLoader.findClass (URLClassLoader.java:476)
    at java.lang.ClassLoader.loadClass (ClassLoader.java:589)
    at java.lang.ClassLoader.loadClass (ClassLoader.java:522)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:270)
    at java.lang.Thread.run (Thread.java:829)

```

**Comment by pabloem** [comment]
hm those are helper classes that the original dev added, but I'm not using them much. Couldyou focus on the SyndeoTemplate class and its tests?

**Comment by pabloem** [comment]
r: @damondouglas ?

**Comment by pabloem** [comment]
oops let me fix the tests...

**Review by damondouglas — CHANGES_REQUESTED** [review]

**Comment by pabloem** [comment]
@damondouglas sorry about the troubles - I've amended the README.

File: syndeo-template/README.md [hunk 5]
```diff
@@ -18,46 +18,90 @@ transforms. Then, the template will take the pipeline spec, and
...
 
-Generate SchemaIO configs: This command will traverse the classpath for the template, find all `SchemaTransform`
-subclasses, and generate a file with the protocol buffer configuration for all of these subclasses.
+### Format code
+
+To apply spotless formatting rules to the Syndeo template code, run the following command:
 
 ```shell
-mvn compile exec:java -Dexec.mainClass="com.google.cloud.syndeo.ConfigGen"
+mvn -B spotless:apply compile -f  unified-templates.xml -pl syndeo-template/pom.xml
```
[the remark this answers is not in the document]
Comment (pabloem) [reply syndeo-template/README.md]: done.

File: syndeo-template/README.md [hunk 6]
```diff
@@ -18,46 +18,95 @@ transforms. Then, the template will take the pipeline spec, and
...
 ```shell
 # Push Template to GCP:
-gcloud dataflow flex-template build gs://$GCS_BUCKET/template.json \
-    --sdk-language "JAVA"  --flex-template-base-image JAVA11 \
-    --image-gcr-path='gcr.io/$GCP_PROJECT/template:latest' \
-    --jar "target/syndeo-template-1.0-SNAPSHOT.jar" \
+gcloud dataflow flex-template build gs://$GCS_BUCKET_NAME/syndeo-template.json  \
+    --metadata-file syndeo-template/metadata.json   --sdk-language "JAVA"  \
+    --flex-template-base-image JAVA11     --image-gcr-path=gcr.io/$GCP_PROJECT/syndeo-template:latest \
+    --jar "syndeo-template/target/syndeo-template-1.0-SNAPSHOT.jar"  \
```
Comment (damondouglas) [inline syndeo-template/README.md:110]: Running this command yielded an error about `No such file or directory: syndeo-template/target/syndeo-template-1.0-SNAPSHOT.jar` which I tried to solve by running `mvn package` but this led to `bigquery-to-bigtable: Missing target image parameter, perhaps you should add a <to><image> configuration parameter to your pom.xml or set the parameter via the commandline (e.g. 'mvn compile jib:build -Dimage=<your image name>')`.

File: syndeo-template/README.md [hunk 7]
```diff
@@ -18,46 +18,95 @@ transforms. Then, the template will take the pipeline spec, and
...
+gsutil mb gs://$TEMP_LOCATION_BUCKET
+```
+
+5. Run the end-to-end test.
+
+The below command tests the workflow by publishing some data to BQ and checking that it makes it over to Bigtable.
+
+
+```shell
+mvn clean package test -f pom.xml -pl syndeo-template/pom.xml  \
```
Comment (damondouglas) [inline syndeo-template/README.md:92]: Everything in this README worked up until this point.  The fork/branch was cloned/checked out in the Google Cloud Shell environment.

```
git log --oneline
290b65a1 (HEAD -> bq-2-bt, origin/bq-2-bt) found right series of incantations
```

This command resulted in:
<details>
<summary>java.lang.ExceptionInInitializerError</summary>

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.google.cloud.syndeo.it.BigTableWriteIT
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.558 sec <<< FAILURE!
testBigQueryToBigTableSmallNonTemplateJob(com.google.cloud.syndeo.it.BigTableWriteIT)  Time elapsed: 3.478 sec  <<< ERROR!
java.lang.ExceptionInInitializerError
        at com.google.api.services.bigquery.Bigquery$Builder.build(Bigquery.java:5251)
        at org.apache.beam.sdk.io.gcp.bigquery.BigQueryServicesImpl$DatasetServiceImpl.<init>(BigQueryServicesImpl.java:594)
        at org.apache.beam.sdk.io.gcp.bigquery.BigQueryServicesImpl$DatasetServiceImpl.<init>(BigQueryServicesImpl.java:537)
        at org.apache.beam.sdk.io.gcp.bigquery.BigQueryServicesImpl.getDatasetService(BigQueryServicesImpl.java:190)
        at org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO$Write.validate(BigQueryIO.java:2650)
        at org.apache.beam.sdk.transforms.PTransform.validate(PTransform.java:174)
        at org.apache.beam.sdk.Pipeline$ValidateVisitor.enterCompositeTransform(Pipeline.java:662)
        at org.apache.beam.sdk.runners.TransformHierarchy$Node.visit(TransformHierarchy.java:581)
        at org.apache.beam.sdk.runners.TransformHierarchy$Node.visit(TransformHierarchy.java:585)
        at org.apache.beam.sdk.runners.TransformHierarchy$Node.access$500(TransformHierarchy.java:240)
        at org.apache.beam.sdk.runners.TransformHierarchy.visit(TransformHierarchy.java:214)
        at org.apache.beam.sdk.Pipeline.traverseTopologically(Pipeline.java:469)
        at org.apache.beam.sdk.Pipeline.validate(Pipeline.java:598)
        at org.apache.beam.sdk.Pipeline.run(Pipeline.java:322)
        at org.apache.beam.sdk.Pipeline.run(Pipeline.java:309)
        at com.google.cloud.syndeo.transforms.BigTableSchemaTransformTest.setUpBigQueryResources(BigTableSchemaTransformTest.java:128)
        at com.google.cloud.syndeo.it.BigTableWriteIT.testBigQueryToBigTableSmallNonTemplateJob(BigTableWriteIT.java:58)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:242)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:137)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
Caused by: java.lang.IllegalStateException: You are currently running with version 2.0.0 of google-api-client. You need at least version 1.15 of google-api-client to run version 1.25.0 of the BigQuery API library.
        at com.google.common.base.Preconditions.checkState(Preconditions.java:534)
        at com.google.api.client.util.Preconditions.checkState(Preconditions.java:113)
        at com.google.api.services.bigquery.Bigquery.<clinit>(Bigquery.java:48)
        ... 50 more


Results :

Tests in error:
  testBigQueryToBigTableSmallNonTemplateJob(com.google.cloud.syndeo.it.BigTableWriteIT)

Tests run: 1, Failures: 0, Errors: 1, Skipped: 0

```
</details>

**Review by damondouglas — COMMENTED** [review]
@pabloem I don't want to hold up this PR.  If you feel my recent comments are irrelevant to typical users of this repo than feel free to ignore.

File: syndeo-template/README.md (hunk 7 above)
Comment (pabloem) [reply syndeo-template/README.md:92]: this issue did not repro for me. Maybe we can leave this one as is? wdyt?

File: syndeo-template/README.md (hunk 6 above)
Comment (pabloem) [reply syndeo-template/README.md:112]: I reproduced this issue. Thanks Damon! I added the proper package command.

**Pull request merged**